src/stream/src/executor/backfill/no_shuffle_backfill.rs

```diff
@@ -434,64 +422,91 @@ where
                 )
                 .await?;
 
+                tracing::trace!(
+                    epoch = ?barrier.epoch,
+                    ?current_pos,
+                    total_snapshot_processed_rows,
+                    "Backfill state persisted"
+                );
+
                 yield Message::Barrier(barrier);
 
                 // We will switch snapshot at the start of the next iteration of the backfill loop.
             }
         }
 
-        tracing::trace!(
-            actor = self.actor_id,
-            "Backfill has already finished and forward messages directly to the downstream"
-        );
+        tracing::trace!("Backfill has finished, waiting for barrier");
 
         // Wait for first barrier to come after backfill is finished.
         // So we can update our progress + persist the status.
         while let Some(Ok(msg)) = upstream.next().await {
             if let Some(msg) = mapping_message(msg, &self.output_indices) {
                 // If not finished then we need to update state, otherwise no need.
-                if let Message::Barrier(barrier) = &msg && !is_finished {
-                    // If snapshot was empty, we do not need to backfill,
-                    // but we still need to persist the finished state.
-                    // We currently persist it on the second barrier here rather than first.
-                    // This is because we can't update state table in first epoch,
-                    // since it expects to have been initialized in previous epoch
-                    // (there's no epoch before the first epoch).
-                    if is_snapshot_empty {
-                        current_pos =
-                            Some(construct_initial_finished_state(pk_in_output_indices.len()))
-                    }
+                if let Message::Barrier(barrier) = &msg {
+                    if is_finished {
+                        // In the event that some actors already finished,
+                        // on recovery we want to recover the backfill progress for them too,
+                        // so that we can provide an accurate estimate.
+                        // so we call that here, before finally completing the backfill progress.
+                        self.progress.update(
+                            barrier.epoch.curr,
+                            snapshot_read_epoch,
+                            total_snapshot_processed_rows,
+                        );
```

Comment:
Progress is updated here but immediately marked as finished later.

Reply:
This is intentional.
Consider the following case:
1. Some partitions of the stream job are finished (A), and some other partitions are in progress (B)
2. Cluster is restarted.
3. After restart, the progress from mview tracker is only recovered from the backfill executor's persistent state.
4. So if we don't call `update` for A, the total backfill rows could be significantly less than the actual progress.

Reply:
I will add a docstring for this.

Reply:
I see. FYI this update just modifies the `ChainState` in the CN. The later `finish` will replace it into a `Done` state.

Reply:
>  so from meta's view, it will see Done after recovery. I think we should change Done into Done(ConsumedRows) ?

Yeah I guess we can also change `Done` to `Done(ConsumedRows)`, so we can do it in 1 update call. Will update it.

Reply:
https://github.com/risingwavelabs/risingwave/pull/12493/commits/cd4e3d55501285e500fa05e9df1f330b8a3efbef